src/naive/time/mod.rs

```diff
@@ -657,6 +815,134 @@ impl NaiveTime {
+    /// assert_eq!(since(from_hmsm(3, 5, 7, 900), from_hmsm(3, 5, 0, 900)),
+    ///            TimeDelta::Forwards(Duration::from_secs(7)));
+    /// assert_eq!(since(from_hmsm(3, 5, 7, 900), from_hmsm(3, 0, 7, 900)),
+    ///            TimeDelta::Forwards(Duration::from_secs(5 * 60)));
+    /// assert_eq!(since(from_hmsm(3, 5, 7, 900), from_hmsm(0, 5, 7, 900)),
+    ///            TimeDelta::Forwards(Duration::from_secs(3 * 3600)));
+    /// assert_eq!(since(from_hmsm(3, 5, 7, 900), from_hmsm(4, 5, 7, 900)),
+    ///            TimeDelta::Backwards(Duration::from_secs(3600)));
+    /// assert_eq!(since(from_hmsm(3, 5, 7, 900), from_hmsm(2, 4, 6, 800)),
+    ///            TimeDelta::Forwards(Duration::from_secs(3600 + 60 + 1) + Duration::from_millis(100)));
+    /// ```
+    ///
+    /// Leap seconds are handled, but the subtraction assumes that
+    /// there were no other leap seconds happened.
+    ///
+    /// ```
+    /// # use chrono::{TimeDelta, NaiveTime};
+    /// # use core::time::Duration;
+    /// # let from_hmsm = NaiveTime::from_hms_milli;
+    /// # let since = NaiveTime::duration_since;
+    /// assert_eq!(since(from_hmsm(3, 0, 59, 1_000), from_hmsm(3, 0, 59, 0)),
+    ///            TimeDelta::Forwards(Duration::from_secs(1)));
+    /// assert_eq!(since(from_hmsm(3, 0, 59, 1_500), from_hmsm(3, 0, 59, 0)),
+    ///            TimeDelta::Forwards(Duration::from_millis(1500)));
+    /// assert_eq!(since(from_hmsm(3, 0, 59, 1_000), from_hmsm(3, 0, 0, 0)),
+    ///            TimeDelta::Forwards(Duration::from_secs(60)));
+    /// assert_eq!(since(from_hmsm(3, 0, 0, 0), from_hmsm(2, 59, 59, 1_000)),
+    ///            TimeDelta::Forwards(Duration::from_secs(1)));
+    /// assert_eq!(since(from_hmsm(3, 0, 59, 1_000), from_hmsm(2, 59, 59, 1_000)),
+    ///            TimeDelta::Forwards(Duration::from_secs(61)));
+    /// ```
+    pub fn duration_since(self, rhs: NaiveTime) -> TimeDelta {
+        //     |    |    :leap|    |    |    |    |    |    |    :leap|    |
+        //     |    |    :    |    |    |    |    |    |    |    :    |    |
+        // ----+----+-----*---+----+----+----+----+----+----+-------*-+----+----
+        //          |   `rhs` |                             |    `self`
+        //          |======================================>|       |
+        //          |     |  `self.secs - rhs.secs`         |`self.frac`
+        //          |====>|   |                             |======>|
+        //      `rhs.frac`|========================================>|
+        //          |     |   |        `self - rhs`         |       |
+
+        // importantly, here if the `secs` are larger, then that `NaiveTime` is larger,
+        // no matter what the value of the frac is.
+        // this has the effect that if the smaller time is in a leap second, we can simply remove `SECOND_AS_NANOS`
+        // from the frac, and then calculate as normal
+
+        match self.secs.cmp(&rhs.secs) {
```

Comment:
Why is this related to the `TimeDelta` work?

Reply:
This used to rely on the fact that the duration could potentially be negative or positive, now that it can't and so the situation is a bit more  complex and the signed integers are harder to work with.

An option that works which is closer to the original implementation is:

```rust
let mut secs = i64::from(self.secs) - i64::from(rhs.secs);
let mut frac = i64::from(self.frac) - i64::from(rhs.frac);

// `secs` may contain a leap second yet to be counted
let adjust = match self.secs.cmp(&rhs.secs) {
    Ordering::Greater => {
        if rhs.frac >= 1_000_000_000 {
            1
        } else {
            0
        }
    }
    Ordering::Equal => 0,
    Ordering::Less => {
        if self.frac >= 1_000_000_000 {
            -1
        } else {
            0
        }
    }
};

secs += adjust;

while frac < 0 {
    frac += 1_000_000_000;
    secs -= 1;
}
match secs.cmp(&0) {
    Ordering::Equal => {
        TimeDelta::Forwards(Duration::new(0, u32::try_from(frac).unwrap()))
    }
    Ordering::Greater => {
        TimeDelta::Forwards(Duration::new(u64::try_from(secs).unwrap(), u32::try_from(frac).unwrap()))
    }
    Ordering::Less => {
        TimeDelta::Backwards(Duration::new(u64::try_from(secs.abs()).unwrap(), u32::try_from(frac.abs()).unwrap()))
    }
}
```

let me know which you think is better? One benefit of the other way is that it doesn't have any `unwrap`s